### PR TITLE
[SPARK-35777][SQL][TESTS] Check all year-month interval types in UDF

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Check all year-month interval types in UDF.

### Why are the changes needed?
New checks should improve test coverage.

### Does this PR introduce _any_ user-facing change?
Yes but `YearMonthIntervalType` has not been released yet.

### How was this patch tested?
Existed UT.

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
@MaxGekk  Since udf result Period still encoded as default YearMonthIntervalType, so we just check it as input is ok. right?

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by MaxGekk** [comment]
@AngersZhuuuu Add the `[TESTS]` tag to PR's title because this is supposed to be about testing only.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -175,6 +175,8 @@ object Cast {
 
     case (from: UserDefinedType[_], to: UserDefinedType[_]) if to.acceptsType(from) => true
 
+    case (_: YearMonthIntervalType, _: YearMonthIntervalType) => true
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:178]: Could you open a separate PR for this, and write test only casts. It seems we need the same for the `DayTimeIntervalType`. Please, do that in the same PR.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:178]: > Could you open a separate PR for this, and write test only casts. It seems we need the same for the `DayTimeIntervalType`. Please, do that in the same PR.

Done in https://github.com/apache/spark/pull/33035

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:178]: @AngersZhuuuu Thank you.

**Comment by MaxGekk** [comment]
Let's merge https://github.com/apache/spark/pull/33035 before this, and come back to the PR later.

**Comment by AngersZhuuuu** [comment]
FYI @MaxGekk merge lasted master yet

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 2]
```diff
@@ -904,25 +905,43 @@ class UDFSuite extends QueryTest with SharedSparkSession {
 
   test("SPARK-34663: using java.time.Period in UDF") {
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:906]: Please, add SPARK-35777 to the test title.

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 3]
```diff
@@ -904,25 +905,43 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
     // Regular case
-    val input = Seq(java.time.Period.ofMonths(11)).toDF("p")
+    val input = Seq(java.time.Period.ofMonths(13)).toDF("p")
+      .select($"p", $"p".cast(YearMonthIntervalType(YEAR, YEAR)).as("p_y"),
+        $"p".cast(YearMonthIntervalType(MONTH, MONTH)).as("p_m"))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:910]: Use `YearMonthIntervalType(MONTH)`

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 4]
```diff
@@ -904,25 +905,43 @@ class UDFSuite extends QueryTest with SharedSparkSession {
 
   test("SPARK-34663: using java.time.Period in UDF") {
     // Regular case
-    val input = Seq(java.time.Period.ofMonths(11)).toDF("p")
+    val input = Seq(java.time.Period.ofMonths(13)).toDF("p")
+      .select($"p", $"p".cast(YearMonthIntervalType(YEAR, YEAR)).as("p_y"),
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:909]: Please, simplify `YearMonthIntervalType(YEAR, YEAR)` to `YearMonthIntervalType(YEAR)`

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 5]
```diff
@@ -904,25 +905,43 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
+    checkAnswer(nullResult, Row(null, null, null) :: Nil)
+    assert(nullResult.schema === new StructType()
+      .add("null_p", YearMonthIntervalType(YEAR, MONTH))
+      .add("null_p_y", YearMonthIntervalType(YEAR, MONTH))
+      .add("null_p_m", YearMonthIntervalType(YEAR, MONTH)))
     // Input parameter of UDF is null
     val nullInput = Seq(null.asInstanceOf[java.time.Period]).toDF("null_p")
+      .select($"null_p",
+        $"null_p".cast(YearMonthIntervalType(YEAR, YEAR)).as("null_p_y"),
+        $"null_p".cast(YearMonthIntervalType(MONTH, MONTH)).as("null_p_m"))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:935]:
```suggestion
        $"null_p".cast(YearMonthIntervalType(YEAR)).as("null_p_y"),
        $"null_p".cast(YearMonthIntervalType(MONTH)).as("null_p_m"))
```

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 6]
```diff
@@ -904,25 +905,43 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
+    val result = input.select(incMonth($"p").as("new_p"),
+      incMonth($"p_y").as("new_p_y"),
+      incMonth($"p_m").as("new_p_m"))
+    checkAnswer(result, Row(java.time.Period.ofMonths(14).normalized(),
+      java.time.Period.ofMonths(13).normalized(),
+      java.time.Period.ofMonths(14).normalized()) :: Nil)
+    assert(result.schema === new StructType()
+      .add("new_p", YearMonthIntervalType(YEAR, MONTH))
+      .add("new_p_y", YearMonthIntervalType(YEAR, MONTH))
+      .add("new_p_m", YearMonthIntervalType(YEAR, MONTH)))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:921]: Just the default one:
```suggestion
      .add("new_p", YearMonthIntervalType())
      .add("new_p_y", YearMonthIntervalType())
      .add("new_p_m", YearMonthIntervalType()))
```

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 7]
```diff
@@ -904,25 +905,43 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
-    checkAnswer(nullResult, Row(null) :: Nil)
-    // TODO(SPARK-35777): Check all year-month interval types in UDF
-    assert(nullResult.schema === new StructType().add("null_p", YearMonthIntervalType()))
+    val nullResult = input.select(nullFunc($"p").as("null_p"),
+      nullFunc($"p_y").as("null_p_y"), nullFunc($"p_m").as("null_p_m"))
+    checkAnswer(nullResult, Row(null, null, null) :: Nil)
+    assert(nullResult.schema === new StructType()
+      .add("null_p", YearMonthIntervalType(YEAR, MONTH))
+      .add("null_p_y", YearMonthIntervalType(YEAR, MONTH))
+      .add("null_p_m", YearMonthIntervalType(YEAR, MONTH)))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:930]: Let's simplify this:
```suggestion
      .add("null_p", YearMonthIntervalType())
      .add("null_p_y", YearMonthIntervalType())
      .add("null_p_m", YearMonthIntervalType()))
```

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 8]
```diff
@@ -904,25 +905,43 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
-    // TODO(SPARK-35777): Check all year-month interval types in UDF
-    assert(constResult.schema === new StructType().add("10_years", YearMonthIntervalType()))
+    val constResult = nullInput.select(constPeriod($"null_p").as("10_years"),
+      constPeriod($"null_p_y").as("p_y_10_years"), constPeriod($"null_p_m").as("pm_10_years"))
+    checkAnswer(constResult, Row(java.time.Period.ofYears(10),
+      java.time.Period.ofYears(10), java.time.Period.ofYears(10)) :: Nil)
+    assert(constResult.schema === new StructType()
+      .add("10_years", YearMonthIntervalType())
+      .add("p_y_10_years", YearMonthIntervalType(YEAR, MONTH))
+      .add("pm_10_years", YearMonthIntervalType(YEAR, MONTH)))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:944]: Let's be consistent with above line:
```suggestion
      .add("p_y_10_years", YearMonthIntervalType())
      .add("pm_10_years", YearMonthIntervalType()))
```

**Review by MaxGekk — COMMENTED** [review]
LGTM in general except of a few comments.

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:906]: DOne

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 8 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 7 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: DOne

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 6 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 5 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: DOne

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 4 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: DOne

**Review by MaxGekk — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master.
Thank you, @AngersZhuuuu .

**Pull request closed without merging**